### PR TITLE
Nautilus

### PR DESCRIPTION
**Pull request by zjia03**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -251,7 +251,7 @@ if(WITH_GSSAPI)
   set(HAVE_GSSAPI ${GSSApi_FOUND})
 endif()
 
-option(WITH_FUSE "Fuse is here" ON)
+option(WITH_FUSE "Fuse is here" OFF)
```
Comment (hanfengzhe-hi) [inline CMakeLists.txt]: WITH_FUSE should be "ON"

**Comment by zjia03** [comment]
new

File: src/osd/CMakeLists.txt [hunk 2]
```diff
@@ -48,7 +48,7 @@ target_link_libraries(osd
   PUBLIC dmclock::dmclock
   PRIVATE
   ${LEVELDB_LIBRARIES}
-  heap_profiler cpu_profiler ${CMAKE_DL_LIBS})
+  heap_profiler cpu_profiler ${CMAKE_DL_LIBS} hw_ec_partial)
```
Comment (Zhiwei-Dai) [inline src/osd/CMakeLists.txt:51]: 应该相应调整为kps_ec吧

Comment (Zhiwei-Dai) [inline src/osd/CMakeLists.txt:51]: 应该相应调整为kps_ec吧

File: src/osd/ECBackend.cc [hunk 3]
```diff
@@ -1187,11 +1192,33 @@ void ECBackend::handle_sub_read_reply(
 	 ++j, ++req_iter, ++riter) {
       ceph_assert(req_iter != rop.to_read.find(i->first)->second.to_read.end());
       ceph_assert(riter != rop.complete[i->first].returned.end());
-      pair<uint64_t, uint64_t> adjusted =
-	sinfo.aligned_offset_len_to_chunk(
+      if(!partial_read) {
+	pair<uint64_t, uint64_t> adjusted = sinfo.aligned_offset_len_to_chunk(
```
Comment (Zhiwei-Dai) [inline src/osd/ECBackend.cc:1196]: 缩进与对齐

File: src/osd/ECTransaction.cc [hunk 4]
```diff
@@ -61,6 +63,11 @@ void encode_and_write(
 
   for (auto &&i : *transactions) {
     ceph_assert(buffers.count(i.first));
+
+    if(overwrite && (write_sid.find(i.first) == write_sid.end())){
```
Comment (Zhiwei-Dai) [inline src/osd/ECTransaction.cc:67]: 存在强制类型转换，是否有风险

File: src/osd/ECTransaction.h [hunk 5]
```diff
@@ -143,11 +158,21 @@ namespace ECTransaction {
...
 	      head_start, tail_finish - head_start);
+	    ldpp_dout(dpp, 20) << __func__ << ": head_start != tail_finish " << head_start << "!=" << tail_finish << dendl;
 	    if (tail_finish > projected_size)
 	      projected_size = tail_finish;
 	  } else {
 	    ceph_assert(tail_finish <= projected_size);
 	  }
+	   uint64_t hi_head_start = 0, hi_head_len = 0;
+
+	   if(
```
Comment (Zhiwei-Dai) [inline src/osd/ECTransaction.h:169]: 类似的空格细节，可以用vs-code扫描一遍

File: src/osd/ECBackend.cc [hunk 6]
```diff
@@ -1705,44 +1733,82 @@ void ECBackend::do_read_op(ReadOp &op)
...
+	      } 
 	messages[k->first].to_read[i->first].push_back(
 	  boost::make_tuple(
 	    chunk_off_len.first,
 	    chunk_off_len.second,
 	    j->get<2>()));
       }
-      ceph_assert(!need_attrs);
-    }
+       // ceph_assert(!need_attrs);
```
Comment (Zhiwei-Dai) [inline src/osd/ECBackend.cc:1791]: 无效代码删除

File: src/osd/ECBackend.cc [hunk 7]
```diff
@@ -1910,14 +1996,19 @@ bool ECBackend::try_state_to_reads()
     }
   } else {
     op->remote_read = op->plan.to_read;
+    dout(20) << __func__ << ": " <<*op << dendl;
```
Comment (Zhiwei-Dai) [inline src/osd/ECBackend.cc:1999]: 如果代码逻辑走到这里，日志级别不小于20时，会出现重复打印

File: src/osd/ECBackend.h [hunk 8]
```diff
@@ -167,6 +167,27 @@ class ECBackend : public PGBackend {
...
+  {
+    set<int> want_to_read;
+    get_want_to_read_shards(&want_to_read);
+
+    set<int> have;
+    map<shard_id_t, pg_shard_t> shards;
+    set<pg_shard_t> error_shards;
+    get_all_avail_shards(hoid, error_shards, have, shards, false);
+
+    if (includes(have.begin(), have.end(), want_to_read.begin(), want_to_read.end())){
```
Comment (Zhiwei-Dai) [inline src/osd/ECBackend.h:181]: 直接return includes()表达式即可

Comment (hanfengzhe-hi) [reply src/osd/ECBackend.h:181]: 修改。

**Pull request merged**